### ISSUE TITLE
Failed to start cAdvisor inotify_init: too many open files

### ISSUE THREAD

**Issue by mjaow** [kubernetes/kubernetes#58081]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

**What happened**:
There are many pods (59) in my node,when i restart it.I found it never start successfully.It just log:
```
F0110 21:30:25.764493   42304 kubelet.go:1230] Failed to start cAdvisor inotify_init: too many open files
goroutine 356 [running]:
k8s.io/kubernetes/vendor/github.com/golang/glog.stacks(0xc42000e001, 0xc421696000, 0xb09, 0x2710)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:766 +0xa7
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).output(0x4fcdd20, 0xc400000003, 0xc420a54000, 0x4c16a06, 0xa, 0x4ce, 0x0)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:720 +0x1a3
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).printf(0x4fcdd20, 0xc400000003, 0x3458e8c, 0x1b, 0xc4210cdd50, 0x1, 0x1)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:655 +0x14f
k8s.io/kubernetes/vendor/github.com/golang/glog.Fatalf(0x3458e8c, 0x1b, 0xc4210cdd50, 0x1, 0x1)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:1145 +0x67
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules(0xc420e58000)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1230 +0xa4
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).(k8s.io/kubernetes/pkg/kubelet.initializeRuntimeDependentModules)-fm()
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2083 +0x2a
sync.(*Once).Do(0xc420e589d0, 0xc420df5e08)
        /usr/lib/golang/src/sync/once.go:44 +0xbe
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).updateRuntimeUp(0xc420e58000)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2083 +0x33d
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).(k8s.io/kubernetes/pkg/kubelet.updateRuntimeUp)-fm()
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1258 +0x2a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420417d70)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:96 +0x5e
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc420417d70, 0x12a05f200, 0x0, 0x1, 0xc4200a2120)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0xad
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420417d70, 0x12a05f200, 0xc4200a2120)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x4d
created by k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1258 +0x1e9

goroutine 1 [chan receive]:
k8s.io/kubernetes/cmd/kubelet/app.run(0xc4207f8600, 0xc420402000, 0xc400000000, 0xc4207ba270)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:595 +0x4cc
k8s.io/kubernetes/cmd/kubelet/app.Run(0xc4207f8600, 0x0, 0x0, 0x0)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:285 +0x39
main.main()
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/kubelet.go:48 +0x8a

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:2197 +0x1

goroutine 5 [chan receive]:
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x4fcdd20)
        /home/sankuai/code/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x7a
"/opt/logs/kubelet/kubelet.ERROR" [readonly] 1232L, 103325C
```
it seems that cadvisor block when watching new files.Does anybody know how it happened and how to fix it?

After following into code ,I found error happened in  kubernetes/vendor/github.com/google/cadvisor/manager/watcher/raw/raw.go:56
following to the function `common.NewInotifyWatcher()`
```
func NewWatcher() (*Watcher, error) {
	fd, errno := syscall.InotifyInit()
	if fd == -1 {
		return nil, os.NewSyscallError("inotify_init", errno)
	}
	w := &Watcher{
		fd:      fd,
		watches: make(map[string]*watch),
		paths:   make(map[int]string),
		Event:   make(chan *Event),
		Error:   make(chan error),
		done:    make(chan bool, 1),
	}

	go w.readEvents()
	return w, nil
}
```
the code says the syscall InotifyInit return error 'too many open files'

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):v1.6.6
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):CentOS Linux 7 (Core)
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by GitHubxsy** [comment]
+1

**Comment by dims** [comment]
duplicate of https://github.com/kubernetes/kubernetes/issues/10421 ?

**Comment by mjaow** [comment]
@dims it seems that latest k8s with cAdvisor still not solve this issue

**Comment by khenidak** [comment]
The issue is in the underlying inotify implementation i think it does not close the watcher fd correctly https://github.com/golang/exp/commit/292a51b8d262487dab23a588950e8052d63d9113#diff-913c4cd2428ce8671839b1afb2699b25 

I am not sure how can fix it, I can move cAdisor on top of fsnotify

**Comment by dims** [comment]
@khenidak looks like @dashpole is fixing this with https://github.com/google/cadvisor/pull/1916

**Comment by huzhengchuan** [comment]
I thinks the error is still not fix, because google/cadvisor#1959 revert google/cadvisor#1916
